[backend-api] create, update, delete products

simara-svatopluk opened the pull request:
| Q             | A
| ------------- | ---
|Description, reason for the PR| Backend API supports writing operation for products - create, update (even partially), delete
|New feature| Yes
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No <!-- Do not forget to update UPGRADE.md -->
|Fixes issues| ... <!-- Write "closes #123" for the issue to be closed automatically during merge -->
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

simara-svatopluk commented on packages/backend-api/src/Model/Product/ProductApiDataValidator.php:
```diff
@@ -0,0 +1,119 @@
+use Shopsys\FrameworkBundle\Component\Domain\Domain;
+use Symfony\Component\Validator\Constraints;
+
+class ProductApiDataValidator extends AbstractValidator implements ProductApiDataValidatorInterface
```
The `AbstractValidator` shouldn't be abstract and should be kind of `ValidatorRunner` with public method. And should be used here as private `$validatorRunner`

simara-svatopluk commented on packages/backend-api/install/tests/ShopBundle/Smoke/BackendApiCreateProductTest.php:
```diff
@@ -0,0 +1,105 @@
+
+        $expectedResponse = [
+            'code' => 400,
+            'message' => 'During validation of data has been found these errors: [name][xx] This field was not expected., [sellingTo] The value "12.1.2019" is not a valid DateTime::ATOM format.',
```
structure of errors should be different, see https://github.com/shopsys/shopsys/blob/f4a86c067fcc77f3cc057a13122e9803b19299fb/docs/backend-api/api-methods.md#add-product "error response"

simara-svatopluk replied:
So we agreed that `ProductApiDataValidatorInterface::validate` will return an array of errors. And controller will always check error count and if it is `> 0` then it will return a different `View`

simara-svatopluk commented on packages/backend-api/src/Model/Product/ProductDataFactory.php:
```diff
@@ -0,0 +1,62 @@
+use Shopsys\FrameworkBundle\Model\Product\ProductData;
+use Shopsys\FrameworkBundle\Model\Product\ProductDataFactoryInterface as BaseProductDataFactoryInterface;
+
+class ProductDataFactory extends ApiFactory implements ProductDataFactoryInterface
```
for extensibility reasons - when I need to add eg. prices, I have to extend validator, factory but not controller. Interfaces helps with this

grossmannmartin commented on packages/backend-api/install/tests/ShopBundle/Smoke/BackendApiCreateProductTest.php:
```diff
@@ -0,0 +1,187 @@
+
+        $uuid = $this->extractUuid($location);
+
+        $this->getProduct($uuid);
```
Missing assert?

simara-svatopluk replied:
added expecting empty product - product with correct structure, but mostly `null` content

grossmannmartin commented on packages/backend-api/src/Component/DataSetter/ApiDataSetter.php:
```diff
@@ -0,0 +1,95 @@
+     * @param object $entityData
+     * @param string|null $entityDataKey
+     */
+    public function setValueIfExists(string $arrayKey, array $apiData, object $entityData, ?string $entityDataKey = null): void
```
`object` is really wide. I can pass any object I want? 
Couldn't it be narrowed?

simara-svatopluk replied:
Well idea of this `ApiDataSetter` is to be able to set any data object in the future. So it is true that right now it sets only `ProductData`, but in the future it could be `OrderData`, ...

So I can't come with a better typehint

grossmannmartin replied:
Well the interface came in my mind, but I have some feeling that you will not agree 😄

simara-svatopluk replied:
correct

grossmannmartin replied:
okay then :)

grossmannmartin commented on docs/backend-api/api-methods.md:
```diff
@@ -193,19 +197,262 @@ You will need to generate new token after the first one expires.
+    {
+        "message": "Provided data did not pass validation",
+        "errors": [
+            "[sellingTo]" => "The value \"12.1.2019\" is not a valid DateTime::ATOM format.",
```
Why square brackets? Seems not important to me.

simara-svatopluk replied:
That's because the Symfony validator produces it in this form and we didn't have a reason for removing them.

Nested errors have keys for example `[shortDescription][1]`, so "just removing" would be eventually complicated.

Do you think I shall change the format in some way? If so, please suggest one and keep in mind nested errors

grossmannmartin replied:
I liked the format initially introduced in documentation:
```json
{
  "errors": {
    "sellingTo":  "The value \"12.1.2019\" is not a valid DateTime::ATOM format.",
    "shortDescription.1": "Bad word found."
  }
}
```

or nested structure 
```json
{
  "errors": {
    "sellingTo":  "The value \"12.1.2019\" is not a valid DateTime::ATOM format.",
    "shortDescription": {
      "1": "Bad word found."
    }
  }
}
```
and the main argument to handle this: **Returned response is not a valid JSON right now.**

simara-svatopluk replied:
I can only agree. Changed to dot notation

grossmannmartin replied:
I still struggle with the format of the response.
```
{
    "message": "Provided data did not pass validation",
    "errors": [
        "sellingTo" => "The value \"12.1.2019\" is not a valid DateTime::ATOM format.",
        "longDescription.1" => "The value 345 is not a valid string.",
    ]
}
```

is not valid JSON and I believe it should be

simara-svatopluk replied:
how did you get this response?

I just tried to POST with errors, and I got
```json
{
    "code": 400,
    "message": "Provided data did not pass validation",
    "errors": {
        "name.en": "The value \"22\" ......\" cannot be longer then 255 characters",
        "extId": "This field was not expected."
    }
}
```

And this is a valid JSON

grossmannmartin replied:
Ou I didn't test it live... The invalid format is in documentation `api-methods.md` and I assumed it's the same as the return value from the code :)

grossmannmartin commented on docs/backend-api/api-methods.md:
````diff
@@ -193,19 +197,262 @@ You will need to generate new token after the first one expires.
+    **Content**
+    ```
+    {
+        "message": "Product with 74da76db-61ab-43fb-926e-43ba326325fc UUID already exists"
````
According to https://restfulapi.net/http-methods/ PUT method should allow update of the resource.
The PUT method should (have to?) be idempotent, but now multiple identical requests have a different outcome.

simara-svatopluk replied:
You are right!!! And what more, you are right according to the RFC!

We - me, Tom, David - agreed on this solution because we want to provide only one way of updating - PATCH, but still we want to provide a way to create product with specific UUID.
And that's why this situation is documented.

It would be best if we discuss this topic personally (by phone) and come with solution.

simara-svatopluk replied:
so we agreed that UUID will be part of request body and we'll remove PUT method

grossmannmartin commented on docs/backend-api/api-methods.md:
````diff
@@ -193,19 +197,262 @@ You will need to generate new token after the first one expires.
+    }
+    ```
+
+**Update Product**
````
I suggest to name it **Partial Product Update** as it's exactly what it does.

grossmannmartin commented on docs/backend-api/api-methods.md:
````diff
@@ -193,19 +197,262 @@ You will need to generate new token after the first one expires.
+    ```
+    {
+        "code": 400,
+        "message": "This UUID is not valid: f564da76-6d51-4ecd-b004-895c8019a23"
````
More obvious invalid ID would be more helpful. 
```suggestion
        "message": "This UUID is not valid: 123"
```

Right now I was thinking if it's malformed UUID or not existing one in a database

grossmannmartin commented on docs/cookbook/adding-new-attribute-to-an-entity.md:
```diff
@@ -35,12 +34,11 @@ You can find instructions in [Extending Product List](./extending-product-list.m
          * @param \Shopsys\ShopBundle\Model\Product\Product[]|null $variants
          */
-        protected function __construct(BaseProductData $productData, ProductCategoryDomainFactoryInterface $productCategoryDomainFactory, array $variants = null)
+        protected function __construct(BaseProductData $productData, array $variants = null)
```
```suggestion
        protected function __construct(BaseProductData $productData, ?array $variants = null)
```

simara-svatopluk replied:
applied manually

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductApiDataValidator.php:
```diff
@@ -0,0 +1,168 @@
+    {
+        $constraintCollection = new Constraints\Collection($this->getConstraintDefinition());
+
+        return $this->validationRunner->runValidation($productApiData, $constraintCollection);
```
As this method should behave the same as `validateCreate` (according to documentation), I prefer to simplify it to just 
```php
return $this->validateCreate($productApiData);
```

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductApiDataValidator.php:
```diff
@@ -0,0 +1,168 @@
+    /**
+     * @return array
+     */
+    protected function createNameConstrains(): array
```
typo: (https://www.quora.com/What-is-the-difference-between-constraint-and-constrains)
```diff
-    protected function createNameConstrains(): array
+    protected function createNameConstraints(): array
```

Also, method is named createConstraints, but internally working with fields and return value is passed to fields too. Maybe rename it completely could be nice, but I'm not forcing it.

simara-svatopluk replied:
renamed to
* `createNameConstraintDefinition`
* `createDescriptionConstraintDefinition`

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductApiDataValidator.php:
```diff
@@ -0,0 +1,168 @@
+    /**
+     * @return array
+     */
+    protected function createDescriptionConstrains(): array
```
typo: 
```diff
-    protected function createDescriptionConstrains(): array
+    protected function createDescriptionConstraints(): array
```

simara-svatopluk replied:
renamed to
* `createNameConstraintDefinition`
* `createDescriptionConstraintDefinition`

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductApiDataValidatorInterface.php:
```diff
@@ -0,0 +1,21 @@
+<?php
+
```
strict types please

simara-svatopluk replied:
Sir! Yes, sir!

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+use FOS\RestBundle\Controller\Annotations\Patch;
+use FOS\RestBundle\Controller\Annotations\Post;
+use FOS\RestBundle\Controller\Annotations\Put;
+use FOS\RestBundle\Controller\Annotations\QueryParam;
```
I would personally use `FOS\RestBundle\Controller\Annotations` and annotate with `@Annotations\Get("...` But feel free to leave it like it is.

simara-svatopluk replied:
ok

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+    }
+
+    /**
+     * @Post("/products")
```
Get actions have description. Can you fill it for modifying actions too?

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+    /**
+     * @param \Shopsys\FrameworkBundle\Model\Product\Product $product
+     */
+    protected function disallowWorkingWithVariants(Product $product): void
```
Method name suggest you somehow change behavior of import. I suggest `assertProductIsNotVariantType` instead (or so).

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+    /**
+     * @param array $uuids
+     */
+    protected function validateUuids(array $uuids): void
```
I would like to see this method elsewhere, maybe `Validation` namespace so it could be reused

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+
+        if ($uuid !== null) {
+            if (!Uuid::isValid($uuid)) {
+                $errors['[uuid]'] = 'This UUID is not valid: ' . $uuid;
```
Why you use different approach from delete method for checking valid UUID?
Seems little as violation of DRY

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+            return $this->handleView($this->createValidationView($errors));
+        }
+
+        $productData = $this->productDataFactory->createFromApi($request->request->all(), $uuid);
```
I like how `$request->request->all()` is extracted into variable in `patchProductAction` method.
It's clear what this value represents. Can you do it here too, please?

simara-svatopluk replied:
This validation method bothered me for a while, and even Tom wanted to keep it in this fashion I decided to rewrite it to validate only if the UUID is setted... so please check `validateCreatingProductWithDefinedUuid` and beginning of the `createProductAction` method changed to
```php
        if ($uuid) {
            $this->validateCreatingProductWithDefinedUuid($uuid);
        }

        $productApiData = $request->request->all();
        $errors = $this->productApiDataValidator->validateCreate($productApiData);
```

I hate passing optional parameters when the optionality can be solved in the caller, so this is the result I like

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductController.php:
```diff
@@ -0,0 +1,285 @@
+        $this->validateUuids([$uuid]);
+        $productApiData = $request->request->all();
+        $errors = $this->productApiDataValidator->validateUpdate($productApiData);
+        if (count($errors)) {
```
please be explicit with conditions
```suggestion
        if (count($errors) > 0) {
```

simara-svatopluk replied:
Really? I personally find the sorter version better, sort of idiom

grossmannmartin replied:
Well, I like being explicit :) 
But as we don't have any agreement about this topic in our team it's ok.

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductDataFactory.php:
```diff
@@ -0,0 +1,94 @@
+    {
+        $productData = $this->productDataFactory->create();
+        $productData->uuid = $uuid ?: Uuid::uuid4()->toString();
+        $productData->availability = $this->availabilityFacade->getDefaultInStockAvailability();
```
Does this mean any created product will be by default "In stock" without any way how to change it? 

I don't like it. I would like to have a way to set availability (even if it means I need to find proper value elsewhere)

simara-svatopluk replied:
@PetrHeinz told me this is "the default correct behavior". I cannot argue against anything.

I only know that without availability the system breaks, and importing availability isn't in scope of this US

PetrHeinz replied:
This is actually the only possible behavior I can think of, without widening the scope of the US.

grossmannmartin replied:
I know it's more MVP of API but still, without availability, it's not usable to me.
And if I saw properly, handling this is not planed in any future US in the MVP Epic nor in the roadmap plans.

Adding an optional field for availability really widen the scope so much?

simara-svatopluk replied:
Well... I would leave this discussion for BV, what do you think?

simara-svatopluk replied:
or we can involve @LukasHeinz without waiting for BV!

Martin:
> I know it's more MVP of API but still, without availability, it's not usable to me.
And if I saw properly, handling this is not planed in any future US in the MVP Epic nor in the roadmap plans.

Petr:
> This is actually the only possible behavior I can think of, without widening the scope of the US.

Me...
* Handling availability is not part of this US. We cannot always solve everything
* If implementators needs to work with availability immediately, they can implement handling it by themselves
* Unfortunatelly it is more complicated how should be handled... string constrants? Is it enough? What about stocks? What if the implementators extend availabilities by new ones?
* So I vote for not solving availabilities in this US, and create a new US that will solve only them and do the US right after this one

simara-svatopluk replied:
This PR is a first step of the import API, so anyone can use it. But there is more work to do like categories, prices and of course availability. So availability won't be addressed now

grossmannmartin replied:
Availability is literally only required field when you create a product. I appreciate if we push availabilities on the top of things to do in API :)

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductDataFactoryInterface.php:
```diff
@@ -0,0 +1,26 @@
+<?php
+
```
strict types please

grossmannmartin commented on packages/backend-api/src/Resources/config/services.yml:
```diff
@@ -1,8 +1,20 @@
+        public: false
+
+    Shopsys\BackendApiBundle\:
+        resource: '../../**/*{DataSetter,Factory,Transformer,ValidationRunner,Validator}.php'
```
It's mandatory to have ValidationRunner as it's not expected to be more of them and Validator as it still needs to be configured manually (because of the interface)

simara-svatopluk commented on packages/backend-api/install/tests/ShopBundle/Test/OauthTestCase.php:
```diff
@@ -80,18 +82,42 @@ protected function createOauthToken(): string
         ];
-        $client->request($method, $uri, [], [], $headers, $content);
+        $uri = $this->getDomainBaseUrl() . $path;
+        $client->request($method, $uri, [], [], $headers, $content ? json_encode($content) : null);
```
Well technically there is a difference. `null` sends nothing in content `[]` sends `[]` in content. When our API doesn't expect content, then it ignores it, so it doesn't matter.

But I can come with two reasons for `null`
* it's default `Client::request()` value
* We define in the documentation there should be no content

grossmannmartin replied:
I'm not sure we understand each other. 
I'm referring to`$content` in a condition – whether it's intentional to convert empty array in `$content` to null, because passing empty array to method `runOauthRequest` does that.

simara-svatopluk replied:
you are right, it is incorrect. Fixed

simara-svatopluk commented on packages/backend-api/install/tests/ShopBundle/Test/OauthTestCase.php:
```diff
@@ -80,18 +82,42 @@ protected function createOauthToken(): string
+     * @param string $url
+     * @return string
+     */
+    protected function extractUuid(string $url): string
```
TODO: I'll adjust commit message

simara-svatopluk commented on packages/backend-api/easy-coding-standard.yml:
```diff
@@ -23,3 +23,5 @@ parameters:
         Shopsys\CodingStandards\Sniffs\ObjectIsCreatedByFactorySniff:
             - '*/tests/*'
+        ObjectCalisthenics\Sniffs\Files\FunctionLengthSniff:
+            - '*/src/Controller/**/*Validator.php'
```
You are right. Just one correction - it suppresses sniff for all *Validator.php in monorepo. When the monorepo is split, travis checks packages one by one and finds errors

grossmannmartin replied:
I created #1249  to address this issue in the future.

simara-svatopluk commented on docs/cookbook/backend-api/adding-an-attribute-to-product-export.md:
```diff
@@ -5,12 +5,12 @@ This short cookbook describes the steps that you need to do when you want to add
+namespace Shopsys\ShopBundle\Controller\Api\V1\Product;
 
-use Shopsys\BackendApiBundle\Controller\V1\ApiProductTransformer as BaseApiProductTransformer;
+use Shopsys\BackendApiBundle\Controller\V1\Product\ApiProductTransformer as BaseApiProductTransformer;
```
Nice point. What is the reason for creating a new version of API? It is a major change (not addition) in URL or data structure.

Imagine we'll do major changes in data structure eg. removing fields, moving them into subsections. Could we in such case use anything from V1? It's unclear, depends on the change.

So IF there is not major change in data structure in V2, we could use/extract V1 validator; but when there is a major change in data structure, we'll have to write it from scratch (or copy and rewrite)

grossmannmartin replied:
Ok to me :)

simara-svatopluk commented on docs/cookbook/backend-api/adding-an-attribute-to-product-import.md:
```diff
@@ -0,0 +1,73 @@
+Let's say you have added `extId` attribute to `Product` entity following [the cookbook](/docs/cookbook/adding-new-attribute-to-an-entity.md) and you want to include the attribute in the import as well.
+
+## 1. Extend `ProductDataFactory` for the API
+[`ProductDataFactory`](/packages/backend-api/src/Controller/V1/Product/ProductDataFactory.php) is responsible for converting API data to [`ProductData`](/project-base/src/Shopsys/ShopBundle/Model/Product/ProductData.php).
```
changed to `...is responsible for converting request body to ProductData...`

simara-svatopluk commented on docs/upgrade/UPGRADE-unreleased.md:
```diff
@@ -258,5 +258,11 @@ There you can find links to upgrade notes for other versions too.
 ## [shopsys/coding-standards]
 - run `php phing standards-fix` to fix code style as we now disallow Yoda style for comparison in the Shopsys Framework coding standards ([#1209](https://github.com/shopsys/shopsys/pull/1209))
 
+## [shopsys/backend-api]
```
That change isn't trivial - it is involved in multiple files like documentation, maybe configs, maybe tests... It would take I guess 2 hours of devel time to "do it"... Please come with this topic on standup

grossmannmartin replied:
We decided to solve this in separate PR. Resolving conversation here.

simara-svatopluk commented on packages/backend-api/install/app/config/packages/test/fos_rest.yml:
```diff
@@ -0,0 +1,3 @@
+fos_rest:
+    exception:
+        debug: true
```
commit message explains it as
> fos rest bundle is set to be in debug mode in test environment so it returns exceptions correctly during smoke tests

Tom is author of this fix, it took him like 2 hours to figure it out :-)

grossmannmartin replied:
Ou, I missed that. Thanks :)

simara-svatopluk commented on packages/backend-api/src/Component/Validation/ValidationRunner.php:
```diff
@@ -0,0 +1,41 @@
+
+        $errors = [];
+
+        if (count($violations) > 0) {
```
I prefer even `if ($violations->count()) {`

grossmannmartin replied:
Well, I like being explicit :) 
But as we don't have any agreement about this topic in our team it's ok.

grossmannmartin commented on packages/backend-api/install/tests/ShopBundle/Smoke/BackendApiCreateProductTest.php:
```diff
@@ -0,0 +1,215 @@
+        $this->assertSame($uuid, $extractedUuid);
+
+        $foundProduct = $this->getProduct($extractedUuid);
+        unset($product['uuid']);
```
You unset uuid here and also in `getProduct` method. But you can compare products incl. UUIDs, right?

simara-svatopluk replied:
Well, I removed `unset` and had to fix tests with POST without UUID. The outcome is -5 and +5 lines :-D I'll keep it after the change

grossmannmartin commented on packages/backend-api/src/Controller/V1/Product/ProductApiDataValidator.php:
```diff
@@ -0,0 +1,172 @@
+    protected function createDescriptionConstraintDefinition(): array
+    {
+        $descriptionFields = [];
+        foreach ($this->domain->getAllIds() as $locale) {
```
```suggestion
        foreach ($this->domain->getAllIds() as $domainId) {
```
and accordingly in usage below

grossmannmartin commented:
Hi @simara-svatopluk, I noticed the upgrade notes are still present in this PR.
Can you rebase and remove them as they're not needed anymore?

grossmannmartin commented:
Hi, I did user testing.

1. It's mandatory to send header `Content-Type: application/json`, and withou it it just always create empty product no matter what you send in a request body. 
It was completely confusing to me and took long time to figure out what is going on.

2. sellingFrom allows to send datetime, but returned time is always midnight.
```
{
  "name": {
      "en": "22\" Sencor SLE 22F46DM4 HELLO KITTY",
      "cs": "22\" Sencor SLE 22F46DM4 HELLO KITTY"
  },
   "sellingFrom": "2010-01-16T07:05:00+00:00"
} 
```
get response 
```
{
    "uuid": "e99a4960-e96d-47eb-bb53-56c6b8a56c46",
    "name": {
        "en": "22\" Sencor SLE 22F46DM4 HELLO KITTY",
        "cs": "22\" Sencor SLE 22F46DM4 HELLO KITTY"
    },
    "hidden": false,
    "sellingDenied": false,
    "sellingFrom": "2010-01-16T00:00:00+00:00",
    "sellingTo": null,
```


3. PATCH method returns 204 status code, but documentation states it should be 201.

4. in PATCH uuid field is completely ignored even for validation
```
{
	"uuid": "sdfsdf91805aa2bc8",
  "hidden": false
} 
```
But we agreed on this and need for valid uuid which will be ignored in the end would be even more weird. Just wanted to point this out :) no need to do anything

And I'm going to test cookbooks (extensibility)

grossmannmartin commented:
Testing the cookbook:

PATCH method with null value in `extId` field results in 500 exception with following content

```json
{
    "code": 500,
    "message": "An exception occurred while executing 'UPDATE products SET ext_id = ?, recalculate_availability = ?, recalculate_visibility = ? WHERE id = ?' with params [null, 1, 1, 192]:\n\nSQLSTATE[23502]: Not null violation: 7 ERROR:  null value in column \"ext_id\" violates not-null constraint\nDETAIL:  Failing row contains (192, 4, 2, 2, null, 2, null, null, null, , null, , null, null, null, f, f, f, f, f, null, null, t, f, f, t, none, 0, 4c2053a8-acb0-4f21-93dd-357ac2f3d1c6, null)."
}
```

_same occurs when trying to set `"hidden": null`_

simara-svatopluk commented:
1. fixed by checking Content-Type header, added documentation
2. that's true. We just use ATOM format, and it have to contain time; But the field itself is a date, I don't know how to fix this (is it even error?)
3. fixed docs
4. ok, I do nothing
5. fixed by not allowing `null` for `hidden`, `sellingDenied` and `extId` in cookbook

grossmannmartin commented on docs/backend-api/api-methods.md:
```diff
@@ -176,36 +179,255 @@ You will need to generate new token after the first one expires.
+
+  `DELETE`
+
+*  **URL Params**
```
DELETE method also checks for content-type and can return error-response

simara-svatopluk replied:
added documentation (I don't want to hack FOSRest bundle to skip this method)
